File: ThreadsApp/ThreadsApp/Network/ServerAPI.m
//
//  ServerAPI.m
//  ThreadsApp
//
//  Created by Brooma Service on 02/04/2019.
//  Copyright © 2019 Brooma Service. All rights reserved.
//

#import "ServerAPI.h"
#import "ConfigHelper.h"

static NSString* const kAPI = @"api/";
static NSString* const kSignatureEndpoint = @"auth/getSignature/";
static NSString* const kSignatureResponseKey = @"encodedSignature";
static BOOL const kIsDebugLoggingEnabled = YES;

@implementation ServerAPI

+ (void) getClientIdSignature: (NSString*) clientId withCompletion: (THRClientIdSignatureCompletion) completion {
    
    NSString* baseUrl = [ConfigHelper getServerUrl];
    
    if (baseUrl.length) {
        NSString *urlString = [NSString stringWithFormat:@"%@%@%@?clientId=%@",
                               baseUrl, kAPI, kSignatureEndpoint, clientId];
        
        [self loadSignatureWithURL: urlString
                        completion: completion];
        
    } else {
        NSLog(@"Signature loading failed: Server url empty");
    }
}

+ (void) loadSignatureWithURL: (NSString *) url
                   completion: (THRClientIdSignatureCompletion) completion {
    
    NSMutableURLRequest * request = [self request: url];
    
    if (kIsDebugLoggingEnabled) {
        NSLog(@"Load signature: REQUEST = %@", request);
        NSLog(@"Load signature: HEADERS = %@", request.allHTTPHeaderFields);
    }
    
    [NSURLConnection sendAsynchronousRequest: request
                                       queue: [NSOperationQueue currentQueue]
                           completionHandler: 
     ^(NSURLResponse *response, NSData *data, NSError *error) {
         
         if (kIsDebugLoggingEnabled) {
             NSLog(@"Load signature: RESPONSE = %@", response);
             NSLog(@"Load signature: DATA     = %@",
                   data ? [[NSString alloc] initWithData:data
                                                encoding:NSUTF8StringEncoding] : data);
         }
         
         if (error) {
             NSLog(@"Load signature: ERROR    = %@", error);
             completion(nil, error);
         } else {
             NSError *jsonError;
             id json = [NSJSONSerialization JSONObjectWithData:data options: kNilOptions error: &jsonError];
             if (jsonError) {
                 NSLog(@"Load signature: ERROR    = %@", jsonError);
                 completion(nil, jsonError);
             } else {
                 NSString* signature = json[kSignatureResponseKey];
                 if (signature.length) {
                     completion(signature, nil);
                 } else {
                     NSError* signatureDataError = [NSError errorWithDomain:[[NSBundle bundleForClass:[self class]] bundleIdentifier]
                                                                       code:0
                                                                   userInfo:@{@"error_description" : @"clientSignature is absent or empty"}];
                     NSLog(@"Load signature: ERROR    = %@", signatureDataError);
                     completion(nil, signatureDataError);
                 }
             }
         }
     }];
}

+ (NSMutableURLRequest *) request: (NSString *) urlString {
    
    NSURL *url = [NSURL URLWithString: urlString];
    NSMutableURLRequest* signatureRequest = [NSMutableURLRequest requestWithURL: url
                                                                    cachePolicy: NSURLRequestReloadIgnoringLocalAndRemoteCacheData
                                                                timeoutInterval: 30];
    [signatureRequest setHTTPMethod: @"GET"];
    return signatureRequest;
}

@end

File: ThreadsApp/ThreadsAppSwift/Network/ServerAPI.swift
//
//  ServerAPI.swift
//  THRAppSwift-Polarbear
//
//  Created by Brooma Service on 15/05/2019.
//  Copyright © 2019 Brooma Service. All rights reserved.
//

import Foundation

typealias THRClientIdSignatureCompletion = (String?, Error?) -> Void

class ServerAPI {
    
    static let kAPI = "api/"
    static let kSignatureEndpoint = "auth/getSignature/"
    static let kSignatureResponseKey = "encodedSignature"
    static let kIsDebugLoggingEnabled = true
    
    
    class func getSignature(forClientId clientId: String, withCompletion completion: @escaping THRClientIdSignatureCompletion) {
        
        let baseUrl = ConfigHelper.getServerUrl()
        
        if (!(baseUrl ?? "").isEmpty) {
            
            let urlString = String.init(format: "%@%@%@?clientId=%@", baseUrl!, ServerAPI.kAPI, ServerAPI.kSignatureEndpoint, clientId)
            
            if let url = URL(string: urlString) {
                loadSignature(withUrl: url, completion: completion)
            } else {
                print("Signature loading failed: Server url invalid: ", urlString)
            }
            
        } else {
            let signatureLoadError = NSError(domain: Bundle(for: self).bundleIdentifier ?? "",
                                             code: 0,
                                             userInfo: ["error_description": "Signature loading failed: Server url empty"])
            
            print("Load signature: ERROR    = ", signatureLoadError);
            completion(nil, signatureLoadError);
        }
    }
    
    class func loadSignature(withUrl url: URL, completion: @escaping THRClientIdSignatureCompletion) {
        
        var signatureRequest = URLRequest(url: url, cachePolicy: .reloadIgnoringLocalAndRemoteCacheData, timeoutInterval: 2.0)
        signatureRequest.httpMethod = "GET"
        
        
        if (ServerAPI.kIsDebugLoggingEnabled) {
            print("Load signature: REQUEST =", signatureRequest)
            print("Load signature: HEADERS =", signatureRequest.allHTTPHeaderFields ?? "nil")
        }
        
        URLSession.shared.dataTask(with: signatureRequest) { (data, response, error) in
            
            if (ServerAPI.kIsDebugLoggingEnabled) {
                print("Load signature: RESPONSE = ", response ?? "nil");
                print("Load signature: DATA     = ", (data == nil) ? "nil" : String(data: data!, encoding: .utf8) ?? "nil");
            }
            
            if (error != nil) {
                print("Load signature: ERROR    = ", error!)
                completion(nil, error)
            } else {
                
                do {
                    let json = try JSONSerialization.jsonObject(with: data!, options: []) as? Dictionary<String, Any>
                    
                    let signature = json?[ServerAPI.kSignatureResponseKey] as? String
                    
                    if (!(signature ?? "").isEmpty) {
                        completion(signature, nil);
                    } else {

                        let signatureDataError = NSError(domain: Bundle(for: self).bundleIdentifier ?? "",
                                                         code: 0,
                                                         userInfo: ["error_description": "clientSignature is absent or empty"])
                        
                        print("Load signature: ERROR    = ", signatureDataError);
                        completion(nil, signatureDataError);
                    }
                    
                } catch let jsonError {
                    print("Load signature: ERROR    = ", jsonError);
                    completion(nil, jsonError);
                }
            }
        }.resume()
        
    }
    
}
